### DISCUSSION TITLE
Form input losses control

### DISCUSSION THREAD

**Discussion by howdyAnkit** [react-hook-form/react-hook-form#7560] [Q&A] [answered]
When I enter the value in form input against the matching pattern I get an error that's expected behaviour but when I try to enter a value for the second time the form input just loses control after entering the first character please find below-attached csb for more details

https://codesandbox.io/s/inspiring-bardeen-wuf8j?file=/src/App.js

**Comment by Moshyfawn** [comment, marked as answer]
Move your `<InputDiv />` initialization out of the `<App />`'s body or put it inside `useCallback` to fix the issue. See [this comment](https://github.com/react-hook-form/react-hook-form/discussions/7530#discussioncomment-1968716) in #7530 discussion for the explanation.
Working [CSB](https://codesandbox.io/s/elastic-cdn-f9ncf?file=/src/App.js:370-373)
TL;DR the input component gets re-initialized every `<App />` re-render, so RHF looses track of it.

**Reply by howdyAnkit** [reply]
Thanks for this @Moshyfawn was really pulling my hair for 3-4 hrs.

**Reply by Moshyfawn** [reply]
No problem!

**Reply by howdyAnkit** [reply]
Hey @Moshyfawn can you check the same Csb link and let me know why it's having problems when I provide 
``` valueAsNumber: true,``` the other validation stops working but as soon as I remove valueAsNumber every other validation seems to work

**Reply by Moshyfawn** [reply]
Phone numbers **are strings** of digit characters, **not numbers**. You're looking to validate the value's length and pattern, but since you convert the input string to a number by setting `valueAsNumber` prop to `true`, `pattern` and length _(`minLength`, `maxLength`)_ validations don't apply.

I'd suggest leaving phoneNumber input as string to validate it against the pattern and value length.

If you 100% need to convert the string value to number at input, you can instead provide a custom validation by checking the value against it's length and pattern yourself by providing the [`validate` function](https://react-hook-form.com/api/useform/register#main) . But you gonna want to convert it to string to do so anyway, so it doesn't make sense. On top of that, you gonna run into trouble having `NaN` value when user inputs non-numeric characters. It's a can of worms you best avoid by having the phoneNumber input as string.

**Reply by Moshyfawn** [reply]
By the way, you can provide your custom validation error messages directly into the validation props like this:
```
required: "Please, enter your phone number"
pattern: {
  value: /^[0-9]*$/,
  message: ""Please, enter a valid phone number
}
```
So, instead of having a bunch of conditions to display your errors, you can do
```
<p>{errors.phoneNumber?.message}</p>
```

Check out [the Register Options section of RHF docs](https://react-hook-form.com/api/useform/register#main) to learn more
